### PR TITLE
Support passing parameters to a QuarkusTestResource

### LINKED ISSUE #9066: Arguments for QuarkusTestResource
**Description**
We use `@QuarkusTestResource` to start databases using Testcontainers, which works very nicely. It would be nice if the `@QuarkusTestResource` annotation allowed passing arguments to the `QuarkusTestResourceLifecycleManager`; e.g. to allow specifying the tag for the Docker container image.

### PR DESCRIPTION
**Pull request by geoand**
Resolves: #9066

### CODE REVIEW

**Comment by geoand** [comment]
This nees to be to ensure that we don't have conflicting args.

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java [hunk 1]
```diff
@@ -23,6 +23,13 @@
      */
     Class<? extends QuarkusTestResourceLifecycleManager> value();
 
+    /**
+     * @return The arguments to be passes to the {@code QuarkusTestResourceLifecycleManager}
+     *
+     * @see QuarkusTestResourceLifecycleManager#init(String[])
+     */
+    String[] args() default {};
```
Comment (gsmet) [inline test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java:31]: Should it be args or initParameters si that what it does is extra clear? Open subject, really, just asking.

**Review by gsmet — DISMISSED** [review]
Interesting! I added a comment and a question.

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java (hunk 1 above)
Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java:31]: Good idea, I'll do it!

Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java]: Fixed

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java [hunk 2]
```diff
@@ -23,6 +23,13 @@
      */
     Class<? extends QuarkusTestResourceLifecycleManager> value();
 
+    /**
+     * @return The arguments to be passes to the {@code QuarkusTestResourceLifecycleManager}
```
[the remark this answers is not in the document]
Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java]: Fixed

**Comment by famod** [comment]
What happens if you have:
- Test 1 with `@QuarkusTestResource(value=MyMgr.class, initArgs={"foo=true"})`
- Test 2 with `@QuarkusTestResource(value=MyMgr.class, initArgs={"foo=false"})`

or something like that?

**Comment by geoand** [comment]
The PR makes sure that the test fails

**Comment by knutwannheden** [comment]
> The PR makes sure that the test fails

Would an alternative be to instantiate `MyMgr` twice (and call one with `init("foo=true")` and the other with `init("foo=false")`)?

**Comment by geoand** [comment]
> > The PR makes sure that the test fails
> 
> Would an alternative be to instantiate `MyMgr` twice (and call one with `init("foo=true")` and the other with `init("foo=false")`)?

I think that could lead to surpising results. Best keep with the established Quarkus practice.

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java [hunk 3]
```diff
@@ -23,6 +23,13 @@
      */
     Class<? extends QuarkusTestResourceLifecycleManager> value();
 
+    /**
+     * @return The arguments to be passed to the {@code QuarkusTestResourceLifecycleManager}
+     *
+     * @see QuarkusTestResourceLifecycleManager#init(String[])
+     */
+    String[] initArgs() default {};
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java:31]: Wouldn't it make more sense for this to be a key-value type arrangement? You can do this by creating a new @ResourceArg annotation with name+value and using an array of annotations.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

**Comment by knutwannheden** [comment]
> I think that could lead to surpising results. Best keep with the established Quarkus practice.

I see. One would have to create some "dummy" classes in these (hopefully rare) situations. I was for instance thinking of the situation where I need two datasources. Then I would have liked to declare two annotations like `@QuarkusTestResource(PostgresDatabase.class)` (default datasource) and `@QuarkusTestResource(PostgresDatabase.class, initArgs="datasource=other")`. Possibly there could be some flag in the future to allow this kind of usage.

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java (hunk 3 above)
Comment (knutwannheden) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java:31]: Yes, that sounds like a good idea. That way only the values need to be deserialized (if not used as strings). The signature of `init()` could change to `init(Map<String, String> args)`.

Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResource.java]: That's an excellent idea @stuartwdouglas!  I updated the PR to include it.

File: test-framework/common/src/main/java/io/quarkus/test/common/ResourceArg.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Uses to define arguments of for {@code QuarkusTestResource}
+ *
+ * see {@link QuarkusTestResource#initArgs()}
+ */
+@Target(ElementType.METHOD)
```
Comment (knutwannheden) [inline test-framework/common/src/main/java/io/quarkus/test/common/ResourceArg.java:14]: I think this should be declared as `@Target({})`.

File: test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResourceLifecycleManager.java [hunk 5]
```diff
@@ -31,6 +31,18 @@
      */
     void stop();
 
+    /**
+     * Arguments passed to the lifecycle manager before it starts
+     * These arguments are taken from {@code QuarkusTestResource#args()}
```
Comment (knutwannheden) [inline test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResourceLifecycleManager.java:36]: Nitpick: `{@code QuarkusTestResource#initArgs()}`

Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResourceLifecycleManager.java:36]: Ah darn... I thought I had changed all of those :)

File: test-framework/common/src/main/java/io/quarkus/test/common/ResourceArg.java (hunk 4 above)
Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/ResourceArg.java]: Ah yeah, that's definitely better

**Comment by gsmet** [comment]
> I think that could lead to surpising results. Best keep with the established Quarkus practice.

It might be just me but I think I would:
- allow two resources with different initArgs
- disallow having two resources with the exact same initArgs

This way, we have the best of both world: flexibilty and keeping the behavior as is if you don't use initArgs.

WDYT?

**Comment by knutwannheden** [comment]
@gsmet What you propose would also be my preferred solution.

**Comment by geoand** [comment]
I don't know... I am not convinced about having different `initArgs`. I think it would lead to users to believe that they can use different `@QuarkusTestResource` on different classes and that those would be applied per test.

**Comment by famod** [comment]
By the way, what's the status of having individual resources per test class? Anything planned?

The way the annotation can be used and how it is documented (JavaDoc) leaves the impression you can have individual resources per test class.

**Comment by geoand** [comment]
> By the way, what's the status of having individual resources per test class? Anything planned?
> 

Nothing done so far, but hopefully soon :)

> The way the annotation can be used and how it is documented (JavaDoc) leaves the impression you can have individual resources per test class.

Would you like to propose a way for it to be written differently that would cause less confusion?

As I mentioned before, I think that the if we allow `@QuarkusTestResource` with different `initArgs`, it will only add to the confusion.

**Comment by geoand** [comment]
So the question becomes, what should we do about this?

**Comment by stuartwdouglas** [comment]
I think that different args makes sense, e.g. starting two different databases of the same type is definitely a common use case, however I can see how it could be confusing. 

Maybe make this annotation @Repeatable, IMHO this would make it clearer that it can be used to create multiple resources.

**Comment by geoand** [comment]
`@QuarkusTestResource` is already repeatable, so I just lifted the restriction I had imposed on not allowing different `initArgs`. 
I also added a note to the javadoc of `@QuarkusTestResource` to make clarify how it's used.

**Comment by geoand** [comment]
CI failure has nothing to do with this PR:

```
2020-05-11T07:49:51.8096777Z Caused by: java.io.IOException: Could not start process: <EOF>
2020-05-11T07:49:51.8096946Z 	at de.flapdoodle.embed.mongo.AbstractMongoProcess.onAfterProcessStart(AbstractMongoProcess.java:79)
2020-05-11T07:49:51.8097071Z 	at de.flapdoodle.embed.process.runtime.AbstractProcess.<init>(AbstractProcess.java:116)
```

**Comment by geoand** [comment]
@stuartwdouglas, @gsmet WDYT now?

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**